File: 797-rabbits-in-forest/rabbits-in-forest.go
func numRabbits(answers []int) int {
    countMap := make(map[int]int);
	result := 0;

    for _, ans := range answers {
		countMap[ans]++;
	}

    for ans, count := range countMap {
		groupSize := ans + 1;
		numGroups := (count + groupSize - 1) / groupSize;
		result += numGroups * groupSize;
	}

    return result;
}

File: 797-rabbits-in-forest/rabbits-in-forest.java
class Solution {
    public int numRabbits(int[] answers) {
        int[] maxArr = new int[1001];
        int result = 0;

        for (int ans : answers) {
            if (++maxArr[ans] == 1) result += ans + 1;
            if (maxArr[ans] == ans + 1) maxArr[ans] = 0;
        }

        return result;
    }
}